### PR TITLE
Handle inputs and outputs in correct order

### PR DESCRIPTION
**Pull request by martijnbastiaan**
Eleminates the need for prepending singals with dummy elements

Fixes https://github.com/clash-lang/clash-vexriscv/issues/1

# TODO after merging:

- [ ] Open issue w.r.t. parallel tests

### CODE REVIEW

File: clash-vexriscv-sim/src/Utils/Storage.hs [hunk 1]
```diff
@@ -2,17 +2,39 @@
...
+  deepErrorX = errorX
+
+  -- Keys are 'Int's and evaluated to WHNF because this is a strict map. For 'Int's,
+  -- WHNF ~ NF, so we only need to check the values.
+  hasUndefined = any hasUndefined . I.elems . unMappedMemory
+
+  -- Not a product type, so no spine
+  ensureSpine = id
+
+  -- This is a strict map, so we dont need to do anything
```
[the remark this answers is not in the document]
Comment (martijnbastiaan) [reply clash-vexriscv-sim/src/Utils/Storage.hs:36]: Yeah for `BitVector` `NF ~ WHNF`. I'll add a comment!

File: clash-vexriscv/src/VexRiscv.hs [hunk 2]
```diff
@@ -236,48 +198,98 @@ vexRiscv# !_sourcePath !_clk rst0
...
+
+      combInput = COMB_INPUT
+        <$> (boolToBit <$> iBus_ACK)
+        <*> (unpack    <$> iBus_DAT_MISO)
+        <*> (boolToBit <$> iBus_ERR)
+        <*> (boolToBit <$> dBus_ACK)
+        <*> (unpack    <$> dBus_DAT_MISO)
+        <*> (boolToBit <$> dBus_ERR)
+
+      wordCast = fromInteger . toInteger
```
[the remark this answers is not in the document]
Comment (martijnbastiaan) [reply clash-vexriscv/src/VexRiscv.hs:217]: Yes!

File: clash-vexriscv/src/VexRiscv.hs [hunk 3]
```diff
@@ -236,48 +198,98 @@ vexRiscv# !_sourcePath !_clk rst0
...
-        pure $ go step (unsafeFromReset rst0) input
-
-      (unbundle -> (iBusM2S, dBusM2S)) = (<$> output) $ \(Output iBus dBus) -> (iBus, dBus)
-
-      (unbundle -> (iBus_ADR, iBus_DAT_MOSI, iBus_SEL, iBus_CYC, iBus_STB, iBus_WE, iBus_CTI, iBus_BTE)) =
-        (<$> iBusM2S) $ \(WishboneM2S a b c _ e f g h i) -> (a, b, c, e, f, g, h, i)
-
-      (unbundle -> (dBus_ADR, dBus_DAT_MOSI, dBus_SEL, dBus_CYC, dBus_STB, dBus_WE, dBus_CTI, dBus_BTE)) =
-        (<$> dBusM2S) $ \(WishboneM2S a b c _ e f g h i) -> (a, b, c, e, f, g, h, i)
+      (v, initStage1, initStage2, stepRising, stepFalling, _shutDown) = unsafePerformIO vexCPU
```
[the remark this answers is not in the document]
Comment (martijnbastiaan) [reply clash-vexriscv/src/VexRiscv.hs:201]: After having done it, agreed!

File: clash-vexriscv/src/VexRiscv.hs [hunk 4]
```diff
@@ -401,16 +410,49 @@ vexRiscv# !_sourcePath !_clk rst0
...
+        poke combInputFFI combInput
+        vexrInitStage2 vPtr combInputFFI
+
+    {-# NOINLINE stepRising #-}
+    stepRising vPtr fsSinceLastEvent nonCombInput =
+      alloca $ \nonCombInputFFI -> alloca $ \outputFFI -> do
+        poke nonCombInputFFI nonCombInput
+        vexrStepRisingEdge vPtr fsSinceLastEvent nonCombInputFFI outputFFI
+        output <- peek outputFFI
+        pure output
```
Comment (christiaanb) [inline clash-vexriscv/src/VexRiscv.hs:448]:
```suggestion
        peek outputFFI
```

File: clash-vexriscv/src/VexRiscv.hs [hunk 5]
```diff
@@ -401,16 +410,49 @@ vexRiscv# !_sourcePath !_clk rst0
...
-      pure $ outputFromFFI outVal
-    shutDown = vexrShutdown v
-  pure (step, shutDown)
+    {-# NOINLINE initStage1 #-}
+    initStage1 vPtr nonCombInput =
+      alloca $ \nonCombInputFFI -> alloca $ \outputFFI -> do
+        poke nonCombInputFFI nonCombInput
+        vexrInitStage1 vPtr nonCombInputFFI outputFFI
+        output <- peek outputFFI
+        pure output
```
Comment (christiaanb) [inline clash-vexriscv/src/VexRiscv.hs:434]:
```suggestion
        peek outputFFI
```

File: clash-vexriscv/src/VexRiscv.hs [hunk 6]
```diff
@@ -236,48 +198,95 @@ vexRiscv# !_sourcePath !_clk rst0
...
+        out1 <- unsafeInterleaveIO (simCycles ts cncs ccs)
+        pure $ out0 :- out1
+
+      simCycles ((fsSinceLastEvent, Falling) : ts) cncs (cc :- ccs) = do
+        stepFalling v fsSinceLastEvent cc
+        simCycles ts cncs ccs
+
+      simCycles [] _ _ = error "Empty ticks: should never happen"
+
+      output = unsafePerformIO (simInitThenCycles nonCombInput combInput)
```
Comment (christiaanb) [inline clash-vexriscv/src/VexRiscv.hs:250]:
```haskell
output = unsafePerformIO $ do
  (v, initStage1, initStage2, stepRising, stepFalling, _shutDown) <- vexCPU
  let simInitThenCycles = ...
      simCycles = ...
  simInitThenCycles nonCombInput combInput
```

**Review by christiaanb — APPROVED** [review]

**Pull request merged**